File: baekjoon_all/18000+/boj_18511.java
// Solve 2024-01-29

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_18511 {

    static int n, k;
    static int nStrLen;
    static int[] digits;
    static int ans;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        String nStr = st.nextToken();
        nStrLen = nStr.length();
        n = Integer.parseInt(nStr);
        k = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine(), " ");
        digits = new int[k];

        for (int i = 0; i < k; i++) {
            digits[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(digits);
        ans = -1;
        dfs(0, 0);
        System.out.println(ans);
        br.close();
    }

    static boolean dfs(int depth, int num) {
        if (depth == nStrLen) {
            if (num <= n) {
                ans = num;
                return true;
            }

            return false;
        }

        for (int i = k - 1; i >= 0; i--) {
            if (dfs(depth + 1, num * 10 + digits[i])) {
                return true;
            }
        }

        if (num == 0) {
            if (dfs(depth + 1, 0)) {
                return true;
            }
        }

        return false;
    }

}

File: baekjoon_all/18000+/boj_18511.cpp
// Solve 2024-01-29

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int n, k;
int len_str_n;
int digits[3];
int ans = -1;

bool dfs(int depth, int num) {
    if (depth == len_str_n) {
        if (num <= n) {
            ans = num;
            return true;
        }

        return false;
    }

    for (int d : digits) {
        if (dfs(depth + 1, num * 10 + d)) {
            return true;
        }
    }

    if (num == 0) {
        if (dfs(depth + 1, 0)) {
            return true;
        }
    }

    return false;
}

int main() {
    FASTIO;

    cin >> n >> k;
    len_str_n = SIZE(to_string(n));

    for (int i = 0; i < k; i++) {
        cin >> digits[i];
    }

    sort(digits, digits + k, greater<int>());
    dfs(0, 0);
    cout << ans << '\n';

    return 0;
}
